trace/lrumap.go

```diff
@@ -15,23 +15,47 @@
 package trace
 
 import (
-	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/golang/groupcache/lru"
 )
 
+// A simple lru.Cache wrapper that tracks the keys of the current contents and
+// the cumulative number of evicted items.
 type lruMap struct {
-	simpleLruMap *simplelru.LRU
+	keys         map[lru.Key]bool
+	cache        *lru.Cache
 	droppedCount int
 }
 
 func newLruMap(size int) *lruMap {
-	lm := &lruMap{}
-	lm.simpleLruMap, _ = simplelru.NewLRU(size, nil)
+	lm := &lruMap{
+		keys:         make(map[lru.Key]bool),
+		cache:        lru.New(size),
+		droppedCount: 0,
+	}
+	lm.cache.OnEvicted = func(key lru.Key, value interface{}) {
+		delete(lm.keys, key)
+		lm.droppedCount++
+	}
 	return lm
 }
 
-func (lm *lruMap) add(key, value interface{}) {
-	evicted := lm.simpleLruMap.Add(key, value)
-	if evicted {
-		lm.droppedCount++
+func (lm lruMap) Len() int {
```

Comment:
Len shouldn't be exported. Same with Keys, Add and Get.

Reply:
I was treating the exported methods (of the unexported struct) as the API of the lruMap, which isn't explicitly contrary to best practices as far as I know. (Though if that is indeed the case, a style link would be appreciated.) But you of course set the convention here. Done!